common/http/src/client.rs

```diff
@@ -45,145 +30,51 @@ impl Client {
-        match self.buffer.as_mut().unwrap(/* poll will panic if this isn't set */).write_line(line).await
-        {
-            Ok(_) => {
-                if let Some(wh) = self.state_write.as_ref() {
-                    if let (Some(key), Some(offset)) = (key.as_ref(), offset) {
-                        debug!("Updating offset for {:?} to {}", key, offset);
-                        wh.update(key, offset).await.unwrap();
-                    }
-                }
-                self.buffer_bytes = self.buffer.as_ref().map(|b| b.bytes_len()).unwrap_or(0);
+                self.make_request(body, offsets.as_deref()).await
             }
-            Err(e) => error!("{:?}", e),
-        }
-    }
-
-    pub fn set_max_buffer_size(&mut self, size: usize) {
-        self.buffer_max_size = size;
-    }
-
-    pub fn set_timeout(&mut self, timeout: Duration) {
-        self.inner.set_timeout(timeout)
-    }
-
-    fn should_flush(&self) -> bool {
-        if self.buffer_bytes >= self.buffer_max_size {
-            trace!("filled buffer, flushing");
-            true
-        } else if self.last_flush.elapsed() > Duration::from_millis(250) {
-            trace!("250ms since last upload, flushing");
-            true
-        } else {
-            false
-        }
-    }
-
-    fn should_retry(&self) -> bool {
-        self.last_retry.elapsed() > self.retry_step_delay
-    }
-
-    async fn flush(&mut self) {
-        if self.buffer.is_none() || self.buffer.as_ref().unwrap().count() == 0 {
-            return;
-        }
+            Err(e) => error!("error polling retry: {}", e),
+            _ => {}
+        };
 
-        let buffer = {
-            match self.buffer_source.next().await {
-                Some(Ok(buf)) => buf,
-                Some(Err(e)) => {
-                    error!("{}", e);
-                    return;
-                }
-                None => {
-                    error!("client buffer stream shut down");
-                    panic!("client buffer stream shut down");
+        //log::warn!("Queueing up offsets");
+        Metrics::http().add_request_size(body.len().try_into().unwrap());
+        if let (Some(wh), Some(offsets)) = (self.state_write.as_ref(), file_offsets) {
+            for (key, offset) in offsets {
+                trace!("Updating offset for {:?} to {}", key, offset);
+                if let Err(e) = wh.update(key, *offset).await {
+                    error!("Unable to write offsets. error: {}", e);
                 }
             }
-        };
-        let buffer = self.buffer.replace(buffer).unwrap();
-        let buffer_size = self.buffer_bytes as u64;
-        self.buffer_bytes = 0;
-        Metrics::http().add_request_size(buffer_size);
-        let body = buffer.end().expect("Failed to close ingest buffer");
-        self.make_request(IngestBodyBuffer::from_buffer(body)).await;
-        self.last_flush = Instant::now();
+        }
+        self.make_request(body, file_offsets.as_deref()).await
```

Comment:
Shouldn't we mark the offsets after the request?

Reply:
This queues up the new offsets, but it doesn't actually get flushed until later in the file after we get a Response::Sent

Reply:
oh! true!